source/simulator/melt.cc

```diff
@@ -1234,6 +1333,272 @@ namespace aspect
   }
 
 
+  namespace
+  {
+    template <int dim>
+    struct PcAssembleData
+    {
+      PcAssembleData (const FiniteElement<dim> &finite_element,
+                      const Quadrature<dim>    &quadrature,
+                      const Mapping<dim>       &mapping,
+                      const UpdateFlags         update_flags,
+                      const unsigned int        n_compositional_fields,
+                      const unsigned int        stokes_dofs_per_cell)
+        :
+        finite_element_values (mapping, finite_element, quadrature,
+                               update_flags),
+        local_dof_indices (finite_element.dofs_per_cell),
+        dof_component_indices(stokes_dofs_per_cell),
+        material_model_inputs(quadrature.size(), n_compositional_fields),
+        material_model_outputs(quadrature.size(), n_compositional_fields)
+      {}
+
+
+      PcAssembleData (const PcAssembleData &scratch)
+        :
+        finite_element_values (scratch.finite_element_values.get_mapping(),
+                               scratch.finite_element_values.get_fe(),
+                               scratch.finite_element_values.get_quadrature(),
+                               scratch.finite_element_values.get_update_flags()),
+        local_dof_indices (scratch.local_dof_indices),
+        dof_component_indices( scratch.dof_component_indices),
+        material_model_inputs(scratch.material_model_inputs),
+        material_model_outputs(scratch.material_model_outputs)
+      {
+
+      }
+
+      virtual ~PcAssembleData ()
+      {}
+
+      FEValues<dim>                            finite_element_values;
+
+      std::vector<types::global_dof_index>     local_dof_indices;
+      std::vector<unsigned int>                dof_component_indices;
+
+      MaterialModel::MaterialModelInputs<dim>  material_model_inputs;
+      MaterialModel::MaterialModelOutputs<dim> material_model_outputs;
+
+    };
+
+    template <int dim>
+    struct PcCopyData
+    {
+      explicit PcCopyData (const unsigned int stokes_dofs_per_cell)
+      {
+        nonzero_dof_indices.reserve(stokes_dofs_per_cell);
+      }
+
+      std::vector<types::global_dof_index>   nonzero_dof_indices;
+    };
+
+    template <int dim>
+    class PcNonZeroAssembler : public SimulatorAccess<dim>
+    {
+      public:
+        PcNonZeroAssembler(IndexSet &nonzero_entries,
+                           std::vector<bool> &is_melt_cell)
+          : nonzero_entries (nonzero_entries),
+            is_melt_cell (is_melt_cell)
+        {}
+
+
+        void
+        local_assemble (const typename DoFHandler<dim>::active_cell_iterator &cell,
```

Comment:
Now that I am looking over this, should I also give this a different name (something like `local_save_nonzero_pc_dofs`)?

Reply:
sure, sounds good